### PR TITLE
[Fix] ZeroRedundancyOptimizer ambiguous error with param groups when pytorch < 1.12.0

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

See issue #778 

## Modification

Add check for param groups and torch version, and give a better instruction to users.

## BC-breaking (Optional)

No

## Use cases (Optional)

Same as before

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/optim/optimizer/zero_optimizer.py [hunk 1]
```diff
@@ -53,6 +57,15 @@ def __init__(self, params, optimizer_type: str, **kwargs):
             '`torch.distributed.optim.ZeroReundancyOptimizer` is only '
             'available when pytorch version >= 1.8.')
         assert is_available(), 'torch.distributed.rpc is not available.'
+        # Avoid the generator becoming empty after the following check
+        params = list(params)
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/zero_optimizer.py:61]: This line can be removed.

**Comment by zhouzaida** [comment]
Please update unit tests in https://github.com/open-mmlab/mmengine/blob/bf5c26538eb977265574e07098c28c8536a07ec9/tests/test_optim/test_optimizer/test_optimizer.py#L727

File: mmengine/optim/optimizer/zero_optimizer.py (hunk 1 above)
Comment (nijkah) [reply mmengine/optim/optimizer/zero_optimizer.py:61]: The comment seems to say "`params` become empty when `params`'s type is a generator after the assert line".

Comment (zhouzaida) [reply mmengine/optim/optimizer/zero_optimizer.py:61]: got it

**Review by nijkah — APPROVED** [review]
LGTM

**Review by zhouzaida — APPROVED** [review]

File: tests/test_optim/test_optimizer/test_optimizer.py [hunk 2]
```diff
@@ -785,6 +774,35 @@ def test_build_zero_redundancy_optimizer(self):
...
+        optim_wrapper_cfg = dict(
+            optimizer=dict(
+                type='ZeroRedundancyOptimizer',
+                optimizer_type='SGD',
+                lr=self.base_lr,
+                weight_decay=self.base_wd,
+                momentum=self.momentum),
+            paramwise_cfg=paramwise_cfg)
+        optim_wrapper = build_optim_wrapper(model, optim_wrapper_cfg)
+        self.assertIsInstance(optim_wrapper.optimizer, ZeroRedundancyOptimizer)
```
Comment (HAOCHENYE) [inline tests/test_optim/test_optimizer/test_optimizer.py:803]: We do not need to assert the type of `optim_wrapper.optimizer` here. Its type is guaranteed by `build_optim_wrapper`

**Review by HAOCHENYE — COMMENTED** [review]
We should also test has the overriden function `state_dict` worked

File: tests/test_optim/test_optimizer/test_optimizer.py [hunk 3]
```diff
@@ -735,28 +735,23 @@ def _check_default_optimizer(self, optimizer, model):
...
+        param_groups = optimizer.param_groups
+        params_set = set(model.parameters())
+        self.assertEqual(
+            sum(len(param_group['params']) for param_group in param_groups),
+            len(params_set))
+        self.assertTrue(
+            all(param in params_set for param_group in param_groups
+                for param in param_group['params']))
+        state_dict = optimizer.state_dict()
+        if torch.distributed.get_rank() == 0:
```
Comment (zhouzaida) [inline tests/test_optim/test_optimizer/test_optimizer.py:747]: using `mmengine.dist.get_rank`

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**